GeomIO/SharedIO.F90

```diff
@@ -188,23 +189,41 @@ subroutine add_vertical_dimensions(bundle, metadata, rc)
       type(ESMF_FieldBundle), intent(in) :: bundle
       type(FileMetaData), intent(inout) :: metadata
       integer, optional, intent(out) :: rc
+
       integer :: status
       integer :: num_levels
       type(StringVector) :: vertical_names
       type(StringVectorIterator) :: iter
-      character(len=:), allocatable :: spec_name, dim_name
-      
-      num_levels = get_num_levels(bundle, _RC)
-      if(num_levels == 0) return
-      vertical_names = get_vertical_dim_spec_names(bundle, _RC)
-      iter = vertical_names%begin()
-      do while(iter /= vertical_names%end())
-         spec_name = iter%of()
-         num_levels = get_vertical_dimension_num_levels(spec_name, num_levels)
-         dim_name = get_vertical_dimension_name(spec_name)
-         call metadata%add_dimension(dim_name, num_levels)
-         call iter%next()
+      character(len=:), allocatable :: dim_name
+      type(VerticalStaggerLoc) :: vert_staggerloc
+      integer :: i, num_vgrid_levels
+      type(ESMF_Field), allocatable :: fieldList(:)
+
+
+      call MAPL_FieldBundleGet(bundle, fieldList=fieldList, _RC)
+
+      vertical_names = StringVector()
+      do i = 1, size(fieldList)
+         _HERE, i, size(fieldList)
+         call MAPL_FieldGet(fieldList(i), vert_staggerloc=vert_staggerloc, _RC)
+         dim_name = vert_staggerloc%get_dimension_name()
+         if (dim_name == "") cycle
+         
+         call MAPL_FieldGet(fieldList(i), num_vgrid_levels=num_vgrid_levels, _RC)
+         call vertical_names%push_back(dim_name)
+         _HERE, i, size(fieldList)
       end do
+
+      associate (e => vertical_names%ftn_end())
+        iter = vertical_names%ftn_begin()
+        do while(iter /= e)
+           call iter%next()
+           dim_name = iter%of()
+           num_levels = vert_staggerloc%get_num_levels(num_vgrid_levels)
```

Comment:
This seems to be where num_vgrid_levels is used, but I don't think that it will have the right value here.